runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/AnnotationLookup.java

```diff
@@ -42,129 +48,130 @@ public class AnnotationLookup {
-		var annotations = NAMED_ANNOTATION2VALUE_GETTER.entrySet();
-		for (Entry<Class<? extends Annotation>, Function<Annotation, String>> entry : annotations) {
-			Class<? extends Annotation> annotationClass = entry.getKey();
-			if (descriptor.hasQualifier(annotationClass)) {
-				Annotation namedAnnotation = descriptor.getQualifier(annotationClass);
-				return entry.getValue().apply(namedAnnotation);
+		Annotation[] qualifiers = descriptor.getQualifiers();
+		if (qualifiers != null) {
+			for (Annotation namedAnnotation : qualifiers) {
+				Class<? extends Annotation> annotationType = namedAnnotation.annotationType();
+				if (NAMED_ANNOTATION_CLASSES.contains(annotationType.getName())) {
+					return namedAnnotationValueGetter(annotationType).apply(namedAnnotation);
+				}
 			}
 		}
 		return null;
 	}
 
-	private static final Map<Class<? extends Annotation>, Function<Annotation, String>> NAMED_ANNOTATION2VALUE_GETTER;
-
-	static {
-		Map<Class<? extends Annotation>, Function<Annotation, String>> annotation2valueGetter = new HashMap<>();
-		annotation2valueGetter.put(jakarta.inject.Named.class, a -> ((jakarta.inject.Named) a).value());
-		loadJavaxClass(
-				() -> annotation2valueGetter.put(javax.inject.Named.class, a -> ((javax.inject.Named) a).value()));
-		NAMED_ANNOTATION2VALUE_GETTER = Map.copyOf(annotation2valueGetter);
+	private static Function<Annotation, String> namedAnnotationValueGetter(
+			Class<? extends Annotation> annotationType) {
+		return NAMED_ANNOTATION2VALUE_GETTER2.computeIfAbsent(annotationType, type -> {
+			try {
+				MethodHandles.Lookup lookup = MethodHandles.lookup();
+				MethodType functionApplySignature = MethodType.methodType(String.class, type);
+				CallSite site = LambdaMetafactory.metafactory(lookup, "apply", //$NON-NLS-1$
+						MethodType.methodType(Function.class), functionApplySignature.erase(),
+						lookup.findVirtual(type, "value", MethodType.methodType(String.class)), //$NON-NLS-1$
+						functionApplySignature);
+				return (Function<Annotation, String>) site.getTarget().invokeExact();
+			} catch (Throwable e) {
+				throw new IllegalStateException(e);
+			}
+		});
 	}
 
-	private static List<Class<?>> getAvailableClasses(Class<?> jakartaClass, Supplier<? extends Class<?>> javaxClass) {
-		List<Class<?>> classes = new ArrayList<>();
-		classes.add(jakartaClass);
-		if (javaxClass != null) {
-			loadJavaxClass(() -> classes.add(javaxClass.get()));
-		}
-		return classes;
+	private static final Map<Class<? extends Annotation>, Function<Annotation, String>> NAMED_ANNOTATION2VALUE_GETTER2 = new ConcurrentHashMap<>();
+	private static final Set<String> NAMED_ANNOTATION_CLASSES = Set.of("jakarta.inject.Named", "javax.inject.Named"); //$NON-NLS-1$//$NON-NLS-2$
+	// TODO: warn about the javax-class?
+
+	private static List<String> getAvailableClasses(String jakartaClass, String javaxClass) {
+		return javaxClass != null && canLoadJavaxClass(javaxClass) //
```

Comment:
why is it needed to check if javax is loadable?

Reply:
This methods checks if the javax-classes are even available in the application in any version and if it's therefore worth to check for them. And, maybe even more important, it emits the warning about javax annotations being 'deprecated'.

But in the past it was asked to improve the implementation because the message is to coarse. So this can definitively improved. OTOH, if support for javax annotation does not require their classes in the build-system we can probably support it much easier and longer and it could be consider to un-deprecate their support.

Reply:
If it can be supported by reflection for very little ongoing maintenance overhead, that seems like a good thing for the consumers.